Add v2.0.5 to Boards Manager JSON file

per1234 opened the pull request:
- Boards Manager URL for testing: https://raw.githubusercontent.com/per1234/HoodLoader2/BM-test/package_NicoHood_HoodLoader2_index.json
  - Using Arduino IDE 1.6.6 if you don't already have the previous version of HoodLoader2 installed via Boards Manager you may need to open Boards Manager twice before the HoodLoader2 entry will appear(https://github.com/arduino/Arduino/issues/3795)
  - If you're still having the problem with the optiboot download failing(https://github.com/arduino/Arduino/issues/3861) then that may cause the IDE to cancel download of the HoodLoader2 JSON file so you may need to temporarily remove the optiboot URL(or maybe just move it to the end of the URL list?).
  - Compile fails with HoodLoader 16u2, etc. using Arduino IDE 1.6.5r5 but I'm guessing this is expected behavior because HoodLoader 2.0.5 requires Arduino IDE 1.6.6.
- You must  create a 2.0.5 release with the installation archive file: https://github.com/per1234/HoodLoader2/releases/download/2.0.5/2.0.5-boards_manager.zip
- I have updated the category value according to the [specification](https://github.com/arduino/Arduino/wiki/Arduino-IDE-1.6.x---package_index.json-format-specification).

I'm happy to make any changes, just let me know.

per1234 committed:
Add v2.0.5 to Boards Manager JSON file

NicoHood commented:
Thanks!

* The board names are wrong. I've updated the names and added 2 new boards.
* The errors are correct, 1.6.6 only
* Lets say I find a bug in this release, is it possible to redo this under the same version number?

per1234 commented:
>The board names are wrong. I've updated the names and added 2 new boards.

Ok, I'll fix that.
>Lets say I find a bug in this release, is it possible to redo this under the same version number?

You can but the problem is that the IDE notifies the user when there is a new version of an installed board so if the version number isn't changed they won't get a notification of the update. In order to install the updated 2.0.5 the previously installed one will need to be removed and then 2.0.5 reinstalled.

NicoHood commented:
Can you also add a note that the user needs to upload the bootloader itself as well?

per1234 commented:
Boards manager doesn't really have a spot for adding notes(though I could add a note under "Boards included in this package). There is a **More Info** link that currently points to https://github.com/NicoHood/HoodLoader2 and I can also add a **Online Help** link to a different URL if you want to add the note to the documentation at either of those locations.

NicoHood commented:
nah thats totally okay. just add the new boards. people should be aware of that. I cannot make everything backwardscompatible and still stable. however the bootloader itself is even backwards compatible

per1234 commented:
Boards names updated:
![clipboard01](https://cloud.githubusercontent.com/assets/8572152/10117589/62f79518-6411-11e5-991c-b5909be4e120.gif)
Note there is a bug in the way Boards Manager displays information that changes between versions: https://github.com/arduino/arduino-ide/issues/1493

NicoHood commented:
Thanks a lot for you help!

The funny thing is you can never add the latest json file to the release since it would change the checksum inside. I also added one commit for a bit release note changes but thats fine and not important. People who use the board manager shouldnt care anyways.

Since I do not want to destroy my test environment, could you please check if that worked all correct?

Question:
Can I add a dependency for IDE 1.6.6? If not, could you please file a bug at the Arduino repo? You know this probably better than me.

per1234 commented:
>Thanks

You're welcome, I'm always happy to add new versions to the JSON files, just let me know.
>People who use the board manager shouldnt care anyways.

I agree, chances are most don't even know where the files are downloaded to so they won't look at the downloaded documentation. They can always click on the More Info link in Boards Manager and read the online information.

I tested it out, installation, compile, and Burn Bootloader/Upload Using Programmer with Uno and Mega 2560 all work fine.

I don't think it's possible to use the current tool dependency system for requiring a certain version of the IDE because it installs the required tools if they aren't present and it couldn't do that with the IDE. I guess a system would need to be implemented where if the IDE doesn't meet the required minimum version then installation would be disabled with a message that it requires an IDE upgrade with a link to the Arduino download page.

NicoHood commented:
Would you like to open a feature request for the last statement? That would really help here since a lot of people might get an update notification now.

per1234 commented:
I made a post in the Arduino forum to see if anyone knows a way to do it that I overlooked: http://forum.arduino.cc/index.php?topic=349752. If I don't find a solution there I'll go ahead with the feature request.

per1234 commented:
pYro_65 suggested that you use preprocessor directives to check for the IDE version:
```c
#if ARDUINO < 10606
#error HoodLoader2 v2.0.5 requires Arduino IDE 1.6.6 or greater. Please update your IDE or use a previous version of HoodLoader2.
```
You could put it in pins_arduino.h if you added the variants that you reference from the Arduino core.

per1234 commented:
Or maybe 1.6.6 is only required for the boards that use the HoodLoader2 variant so you could still reference the others from the Arduino core?

NicoHood commented:
Simple but good.
https://github.com/NicoHood/HoodLoader2/commit/540913108219d408c00a2413d5b472af15cf5ddf

Do you know if I can push a 2.0.5-r2 soon? I do not want to name it 2.0.6. Maybe the release was a bit fast XD

per1234 commented:
>Simple but good.

Great, I'm glad there was a solution!
>Do you know if I can push a 2.0.5-r2 soon

Boards manager considers 2.0.5 to be a higher version than 2.0.5-r2. 2.0.5.1 is considered and invalid.version
![clipboard01](https://cloud.githubusercontent.com/assets/8572152/10131288/051ddf8c-6584-11e5-8e17-6b89c8d05d2c.gif)

NicoHood commented:
2.0.5.2 works? Its somehow a long version number but at least an idea.

Or we simply overwrite it. Hopefully not all people have updated?
I want to add some other features as well, so I want to wait a bit anyways. Or I should bump the bootloader version to 2.1.x

per1234 commented:
>2.0.5.2 works?

No, unfotunately only 3 version numbers are allowed because they use the semver specification. It's up to you how you want to handle the versioning, I'm happy to do the overwrite or whatever when you're ready. I do think that 2.0.5-r2 will cause problems though because Boards Manager automatically installs the highest version unless the earlier version is selected so people may end up getting 2.0.5 and think they have the most recent.

NicoHood commented:
https://github.com/arduino/Arduino/issues/3932

per1234 commented:
@NicoHood can you verify https://github.com/arduino/Arduino/issues/3795 (Need to open Boards Manager twice in 1.6.6 after first adding the new Boards Manager URL before the entry will appear in Boards Manager)? If this is real(which it certainly appears so to me) then I think a note of the issue should be added to the **Installation via Boards Manager** section of the HoodLoader2 wiki now that 1.6.6 has been officially released.

NicoHood commented:
I rather think that you should *push* the github issue. Well the problem is, that 1.6.6 is out now and sealed. Meaning the next fix will arrive in month.

I wanted to try it out, but our firewall here blocks the download. And at home I do not want to mess up my development environment.

Thx for your support.

per1234 commented:
With nobody else bothering to confirm the issue I couldn't know if it was isolated to my system but now I know of one other who has also experienced it so hopefully they'll report to the issue also. I try not to be a nag as I know the developers have lots of things to do but 1.6.6 should never have been released with this unfixed.
